core-to-plc/src/Language/Plutus/CoreToPLC/Plugin.hs

```diff
@@ -0,0 +1,177 @@
+
+-- See Note [Deserializing the AST]
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
+module Language.Plutus.CoreToPLC.Plugin (PlcCode, getSerializedCode, getAst, plugin, plc) where
+
+import           Language.Plutus.CoreToPLC
+import           Language.Plutus.CoreToPLC.Error
+
+import qualified GhcPlugins                      as GHC
+
+import qualified Language.PlutusCore             as PC
+import           Language.PlutusCore.Quote
+
+import           Language.Haskell.TH.Syntax      as TH
+
+import           Control.Monad.Except
+import           Control.Monad.Reader
+import           Control.Exception
+import qualified Data.ByteString.Lazy            as BSL
+import qualified Data.Map                        as Map
+import           Data.Maybe                      (catMaybes)
+import           Data.Text                       as T
+
+-- Note: we construct this by coercing, so this *must* remain representationally equivalent to '[Word]'
+-- unless we change how conversion works, and *must* be abstract.
+-- | A PLC program.
+newtype PlcCode = PlcCode { unPlc :: [Word] }
+
+getSerializedCode :: PlcCode -> BSL.ByteString
+getSerializedCode = BSL.pack . fmap fromIntegral . unPlc
+
+{- Note [Deserializing the AST]
+The types suggest that we can fail to deserialize the AST that we embedded in the program.
+However, we just did it ourselves, so this should be impossible. Possibly we should surface
+the error somehow, but passing it on to the user seems like quite an annoying UI.
+-}
+getAst :: PlcCode -> PC.Program PC.TyName PC.Name ()
+getAst wrapper = let Right p = PC.readProgram $ getSerializedCode wrapper in p
+
+-- | Marks the given expression for conversion to PLC.
+plc :: a -> PlcCode
+-- this constructor is only really there to get rid of the unused warning
+plc _ = PlcCode mustBeReplaced
+
+plugin :: GHC.Plugin
+plugin = GHC.defaultPlugin { GHC.installCoreToDos = install }
+
+install :: [GHC.CommandLineOption] -> [GHC.CoreToDo] -> GHC.CoreM [GHC.CoreToDo]
+install _ todo = pure (GHC.CoreDoPluginPass "C2C" pluginPass : todo)
+
+pluginPass :: GHC.ModGuts -> GHC.CoreM GHC.ModGuts
+pluginPass guts = qqMarkerName >>= \case
+    -- nothing to do
+    Nothing -> pure guts
+    Just name -> GHC.bindsOnlyPass (mapM $ convertMarkedExprsBind name) guts
+
+{- Note [Hooking in the plugin]
+Working out what to process and where to put it is tricky. We are going to turn the result in
+to a String, or maybe a Bytestring or an AST, either way, not the Haskell expression we started with!
+
+Currently we look for calls to the 'plc :: a -> PLC' function, and we replace the whole application with the
+string, which will still be well-typed.
+
+However, if we do this within a single expression, we have problems where GHC gives unconstrained
+type variables the type `Any` rather than leaving them abstracted as we require (see
+note [System FC and system FW]). I think we can do better but I need to figure out how to actually
+look up a name.
```

Comment:
What exactly do you want to look up?

Reply:
I have figured out how to look up names, but I also don't remember what I was getting at here. Not quite verbose enough at the time!